src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -825,86 +835,159 @@ static int parse_imgpath(const std::string &imgpath, Config *cfg)
+  r = parse_args(args, &err_msg, cfg);
+  return r;
+}
 
-  while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
-      break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
+static int get_map_pid(const std::string& pid_path)
+{
+  int pid = 0;
+  std::ifstream ifs;
+  ifs.open(pid_path.c_str(), std::ifstream::in);
+  if (!ifs.is_open()) {
+    return 0;
   }
-
-  close(fd[0]);
-  close(fd[1]);
-
-  return 0;
+  ifs >> pid;
+  return pid;
 }
 
-static int rbd_nbd(int argc, const char *argv[])
+static int do_list_mapped_devices()
 {
   int r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
 
-  Config cfg;
+  std::string default_pool_name;
 
-  enum {
-    None,
-    Connect,
-    Disconnect,
-    List
-  } cmd = None;
+  TextTable tbl;
 
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+ 
   vector<const char*> args;
+  auto cct = global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT,
+                         CODE_ENVIRONMENT_UTILITY, 0);
+ 
+  librados::Rados rados;
+  r = rados.init_with_context(g_ceph_context);
+  if (r < 0) {
+    cerr << "Failed to initialize RADOS: " << cpp_strerror(r) << std::endl;
+    return -1;
+  }
+  r = rados.connect();
+  if (r < 0) {
+    cerr << "Failed to connect to cluster: " << cpp_strerror(r) << std::endl;
+    return -1;
+  }
 
-  argv_to_vec(argc, argv, args);
-  md_config_t().parse_argv(args);
+  r = rados.conf_get("rbd_default_pool", default_pool_name);
+  if (r < 0) {
+    cerr << "Failed to retrieve default pool: " << cpp_strerror(r)
+         << std::endl;
+    return -1;
+  }
+
+  while (true) {
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
+      break;
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
+
+    if(pid > 0) {
+      Config cfg;
+      r = get_mapped_info(pid, &cfg);
+      if (r < 0) {
+        index++;
+        continue;
+      }
+      should_print = true;
+      if (cfg.snapname.empty()) {
+        cfg.snapname = "-";
+      }
+      if (cfg.poolname.empty()) {
+        cfg.poolname = default_pool_name;
+      }
+      tbl << pid << cfg.poolname << cfg.imgname << cfg.snapname
+          << "/dev/nbd" + stringify(index) << TextTable::endrow;
+    }
 
+    index++;
+  }
+
+  if (should_print) {
+    cout << tbl;
+  }
+  return 0;
+}
+
+static int parse_args(vector<const char*>& args, std::ostream *err_msg, Config *cfg)
+{
   std::vector<const char*>::iterator i;
   std::ostringstream err;
 
+  md_config_t().parse_argv(args);
```

Comment:
Instead of this you can do:
```
md_config_t config();
config.parse_config_files(nullptr, nullptr, 0);
config.parse_env();
config.parse_argv(args);
cfg->pool = config.rbd_default_pool;
```

And then `if (cfg->poolname.empty()) ...` code in `do_map()` can be removed as redundant.